File: Python/0002-add-two-numbers.py
# Time:  O(n)
# Space: O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(0)
        cur, carry, = dummy, 0

        while l1 or l2:
            value = carry
            if l1:
                value += l1.val
                l1 = l1.next
            if l2:
                value += l2.val
                l2 = l2.next
            # carry can only me 0 or 1
            carry, value = divmod(value, 10)
            cur.next = ListNode(value)
            cur = cur.next
        if carry == 1:
            cur.next = ListNode(1)

        return dummy.next

File: C++/0002-add-two-numbers.cpp
// Tags: Linked-list Amazon
// Time: O(n)
// Space: O(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummy{0};
        auto cur = &dummy;
        int carry = 0;
        // iterate left to right and carry to right
        while (l1 || l2 || carry) {
            // there may be case when both l1 & l2 are nullptr
            // but carry is 1
            int a = l1 ? l1->val : 0;
            int b = l2 ? l2->val : 0;
            int val = a + b + carry;
            // capture 0th digit and extend linked list
            cur->next = new ListNode(val%10);
            // capture 1's digit
            carry = val/10;
            cur = cur->next;
            l1 = l1 ? l1->next : nullptr;
            l2 = l2 ? l2->next : nullptr;
        }
        return dummy.next;
    }
};
